tests/tests_pytorch/trainer/optimization/test_optimizers.py

```diff
@@ -657,3 +658,67 @@ def lr_scheduler_step(*_): ...
     else:
         with pytest.raises(MisconfigurationException, match="CustomScheduler` doesn't follow"):
             _init_optimizers_and_lr_schedulers(model)
+
+
+@patch("torch.optim.lr_scheduler.StepLR.step")
+def test_lr_scheduler_step_across_epoch_boundaries(mocked_sched, tmp_path):
+    class StepAcrossEpochsModel(LightningModule):
+        def __init__(self):
+            super().__init__()
+            self.layer = torch.nn.Linear(32, 2)
+
+        def forward(self, x):
+            return self.layer(x)
+
+        def training_step(self, batch, batch_idx):
+            # Add print statement to track batch index and global step
+            if hasattr(self, 'trainer'):
+                print(f"Batch idx: {batch_idx}, Global step: {self.trainer.global_step}")
+            return {"loss": torch.tensor(0.1, requires_grad=True)}
+
+        def train_dataloader(self):
+            x = torch.randn(21, 32)
+            y = torch.randn(21, 2)
+            return DataLoader(TensorDataset(x, y), batch_size=3)
+
+        def configure_optimizers(self):
+            optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
+            scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1)
+            return {
+                "optimizer": optimizer,
+                "lr_scheduler": {
+                    "scheduler": scheduler,
+                    "interval": "step",
+                    "frequency": 5,  # Scheduler steps every 5 iterations
+                },
+            }
+
+    model = StepAcrossEpochsModel()
+
+    # Trainer configuration for cross-epoch testing
+    trainer = Trainer(
+        default_root_dir=tmp_path,
+        limit_train_batches=7,  # More than `frequency` iterations per epoch
+        max_epochs=3,  # Test across multiple epochs
+    )
+
+    # Fit the model
+    trainer.fit(model)
+
+    # Debug print statements
+    print(f"Mocked scheduler step calls: {mocked_sched.call_count}")
```

Comment:
Please remove the debug statements, I'd just convert them to asserts that compare the values with expected ones.